spec/abci++/v3.md

```diff
@@ -0,0 +1,202 @@
+decision_p = nil
+lockedValue_p = nil
+validValue_p = nil
+validRound_p = -1
+```
+
+### StartRound(round)
+
+```go
+function startRound(round) {
+    round_p ← round
+    step_p ← propose
+    if proposer(h_p, round_p) = p {
+        if validValue_p != nil {
+            proposal ← validValue_p
+        }
+        else {
+            txdata ← mempool.GetBlock()
+            // getBlockProposal fills in header
+            unpreparedProposal = getUnpreparedBlockProposal(blockdata)
+            proposal = ABCI.PrepareProposal(unpreparedProposal)
+        }
+        broadcast ⟨PROPOSAL, h_p, round_p, proposal, validRound_p⟩
+    }
+    else {
+        schedule OnTimeoutPropose(h_p,round_p) to be executed after timeoutPropose(round_p)
+    }
+}
+```
+
+### ReceiveProposal
+
+In the case where the local node is not locked on any round, the following is ran:
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, −1) from proposer(h_p, round_p) while step_p = propose do {
+  prevote_nil = false
+  if valid(v) ∧ ABCI.VerifyHeader(h_p, v.header) ∧ (lockedRound_p = −1 ∨ lockedValue_p = v) {
+    // We split up the app's header verification from the remainder of its processing of the proposal
+    if ABCI.ProcessProposal(h_p, v).accept {
+      broadcast ⟨PREVOTE, h_p, round_p, id(v,t)⟩ 
+    } else {
+      prevote_nil = true
+    }
+  }
+  else {
+    prevote_nil = true
+  }
+  if prevote_nil {
+    broadcast ⟨PREVOTE, h_p, round_p, nil⟩ 
+    // Include any slashing evidence that may be sent in the process proposal response
+    for evidence in ABCI.ProcessProposal(h_p, v).evidence_list {
+      broadcast ⟨EVIDENCE, evidence⟩ 
+    }
+  }
+  step_p ← prevote
+}
+```
+
+In the case where the node is locked on a round, the following is ran:
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, vr⟩
+ from proposer(h_p, round_p)
+ AND 2f + 1 ⟨PREVOTE, h_p, vr, id(v)⟩ 
+while step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p) do {
+ prevote_nil = false
+  if valid(v) ∧ ABCI.VerifyHeader(h_p, v.header) ∧ (lockedRound_p ≤ vr ∨ lockedValue_p = v) {
+    // We split up the app's header verification from the remainder of its processing of the proposal
+    if ABCI.ProcessProposal(h_p, v).accept {
+      broadcast ⟨PREVOTE, h_p, round_p, id(v)⟩
+    } else {
+      prevote_nil = true
+    }
+  } else {
+   prevote_nil = true
+  }
+  if prevote_nil {
+    broadcast ⟨PREVOTE, h_p, round_p, nil⟩
+    // Include any slashing evidence that may be sent in the process proposal response
+    for evidence in ABCI.ProcessProposal(h_p, v).evidence_list {
+      broadcast ⟨EVIDENCE, evidence⟩ 
+    }
+  }
+  step_p ← prevote
+}
+```
+
+### Prevote timeout
+
+Upon receiving 2f + 1 prevotes, setup a timeout.
+
+```go
+upon 2f + 1 ⟨PREVOTE, h_p, vr, -1⟩ 
+ with step_p = prevote for the first time, do {
+	schedule OnTimeoutPrevote(h_p, round_p) to be executed after timeoutPrevote(round_p)
+}
+```
+
+with OnTimeoutPrevote defined as:
+
+```go
+def OnTimeoutPrevote(height, round) {
```

Comment:
Why not use `function` or `func` everywhere? Or `def` everywhere (I prefer `function`).

Reply:
Agreed, I'll switch to function everywhere